src/MatrixFields/rmul_with_projection.jl

```diff
@@ -0,0 +1,160 @@
+const SingleValue =
+    Union{Number, Geometry.AxisTensor, Geometry.AdjointAxisTensor}
+
+"""
+    mul_with_projection(x, y, lg)
+
+Similar to `x * y`, except that this version automatically projects `y` to avoid
+`DimensionMismatch` errors for `AxisTensor`s. For example, if `x` is a covector
+along the `Covariant3Axis` (e.g., `Covariant3Vector(1)'`), then `y` will be
+projected onto the `Contravariant3Axis`. In general, the first axis of `y` will
+be projected onto the dual of the last axis of `x`.
+"""
+mul_with_projection(x, y, _) = x * y
+mul_with_projection(
+    x::Union{Geometry.AdjointAxisVector, Geometry.Axis2TensorOrAdj},
+    y::Geometry.AxisTensor,
+    lg,
+) = x * Geometry.project(Geometry.dual(axes(x)[2]), y, lg)
+
+"""
+    rmul_with_projection(x, y, lg)
+
+Similar to `rmul(x, y)`, except that this version calls `mul_with_projection`
+instead of `*`.
+"""
+rmul_with_projection(x, y, lg) =
+    rmap((x′, y′) -> mul_with_projection(x′, y′, lg), x, y)
+rmul_with_projection(x::SingleValue, y, lg) =
+    rmap(y′ -> mul_with_projection(x, y′, lg), y)
+rmul_with_projection(x, y::SingleValue, lg) =
+    rmap(x′ -> mul_with_projection(x′, y, lg), x)
+rmul_with_projection(x::SingleValue, y::SingleValue, lg) =
+    mul_with_projection(x, y, lg)
+
+axis_tensor_type(::Type{T}, ::Type{Tuple{A1}}) where {T, A1} =
+    Geometry.AxisVector{T, A1, SVector{length(A1.instance), T}}
+function axis_tensor_type(::Type{T}, ::Type{Tuple{A1, A2}}) where {T, A1, A2}
+    N1 = length(A1.instance)
+    N2 = length(A2.instance)
+    S = SMatrix{N1, N2, T, N1 * N2}
+    return Geometry.Axis2Tensor{T, Tuple{A1, A2}, S}
+end
```

Comment:
Should we use `StaticArraysCore.Size` instead of `A1/A2.instance`? I would even prefer writing a small helper, `sa_len(::Type{SA.SVector{N}}) where {N} = N` to get the length.
Also, it might be nice if we added type restrictions on what `A1`/`A2` are (I assume StaticArrays?)

Reply:
`A1` and `A2` are both of type `Geometry.AbstractAxis`, and the `Geometry` module only has a method for `Base.length(ax::AbstractAxis{I}) where {I} = length(I)`. We could just add a method to `Geometry` for getting the length from the type, like `_length(::Type{<:AbstractAxis{I}}) where {I} = length(I)`?

Reply:
Yeah, I think I'd prefer that